lib/model/typing_status.dart

```diff
@@ -86,3 +89,73 @@ class TypingStatus extends ChangeNotifier {
     }
   }
 }
+
+class TypingNotifier {
+  TypingNotifier({
+    required this.typingStoppedWaitPeriod,
+    required this.typingStartedWaitPeriod,
+  });
+
+  final Duration typingStoppedWaitPeriod;
+  final Duration typingStartedWaitPeriod;
+
+  MessageDestination? _currentDestination;
+  DateTime? _nextStartTime;
+  Timer? _stopTimer;
+
+  void dispose() {
+    _stopTimer?.cancel();
+  }
+
+  void handleTypingStop(PerAccountStore store, MessageDestination destination) {
+    if (destination != _currentDestination) return;
+
+    _notifyStop(store);
+  }
+
+  void handleTyping(PerAccountStore store, MessageDestination destination) {
+    // Two cases.
+    // Renew a previous typing status,
+    // or start a new typing status
+    if (destination == _currentDestination) {
+      _maybeNotifyStart(store);
+      return;
+    }
+
+    _nextStartTime = null;
+    // Start new typing status
+    _currentDestination = destination;
+    _notifyStart(store);
+  }
+
+  Future<void> _maybeNotifyStart(PerAccountStore store) async {
+    if (_nextStartTime == null || DateTime.now().isAfter(_nextStartTime!)) {
+      await _notifyStart(store);
+    }
+  }
+
+  Future<void> _notifyStart(PerAccountStore store) async {
+    _nextStartTime = DateTime.now().add(typingStartedWaitPeriod);
```

Comment:
Quick high-level comment: let's write this so that we definitely don't have to think about time zones to reason about it. :slightly_smiling_face: 

I think an OK version could be made with `DateTime.timestamp()`, which uses UTC. Probably a cleaner approach is to use `Stopwatch` — then we're really not thinking about calendar dates or anything like that, just elapsed durations, which is exactly right for this subsystem.